docs/lsp.md

```diff
@@ -0,0 +1,489 @@
+
+
+## LSP library
+
+The library needs to support all the features we want to implement including the ability to add custom F# ones. It also shouldn't incur an unnecessary performance overhead.
+
+### Candidates
+
+| Library  | Pros | Cons |
+| ---  | --- | --- |
+| We write our own | + Complete control over everything | - A lot of extra work |
+| [Ionide LanguageServerProtocol](https://github.com/ionide/LanguageServerProtocol) | + Battle tested with Ionide and FSAC + Natural F# API | - Might not be able to fully customize it to our need |
+| [Roslyn CLaSP](https://github.com/dotnet/roslyn/blob/main/src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/README.md) | + Should be performant + Future support | - Not stable API - Unnatural C# API |
+| [OmniSharp C# LSP](https://github.com/OmniSharp/csharp-language-server-protocol)  |  | - Unnatural C# API |
+
+It probably makes sense to start with an existing library which should be able to get us to a working product and only then consider writing our own in case it's the only way to enable more functionality or performance.
+
+## LSP client (?)
+
+- Is there any code that can be shared by various clients?
+- Most likely not, since each client will be specific to its platform and project system
+
+
+## Diagnostic mode
+
+Initially we will be offering the **Pull Diagnostics** mode where the client will request diagnostics for a particular document. It will be up to the client to ask for diagnostics for documents at appropriate times and cancel requests that are no longer needed. The client can also use this to implement background checking of all documents in the workspace.
+
+Later we can also add an option to have the server control diagnostics and publish them via **Publish Diagnostics**. Then it will be pushing diagnostics as they are discovered either for active document, open documents or all documents.
+
+## LSIF
+
+We should eventually add support for LSIF. At the moment it looks like we would have to construct it manually (as in, no helpful libraries are available).
+
+This is probably lower priority than anything else.
+
+It might also be a good opportunity to re-think how we store symbols for purposes of finding references, go to definition, or code search (go to symbol in workspace by name/query).
+
+
+## Performance
+
+Areas to pay attention to with regards to performance
+
+- Serialization. Hopefully this will be handled reasonably by a library.
+- Spamming too many requests. Though this will be mainly up to the client, we could think about throttling or debouncing.
+- Caching. This will be mostly up to FCS but we can send hints about what can be cleaned up.
+
+
+## Analyzers & Refactoring support
+
+These will hopefully not require any special treatment and can just use existing LSP APIs.
+
+## Mixed F# / C# solutions
+
+We need to figure out how to talk to C# LSP server for:
+
+- Go to definition
+- Find references / rename
+- (Make sure these work both ways)
```

Comment:
Discussed it with Roslyn team as well, we will make sure it works in both VS and VSCode as soon as we have working LSP server in both.